virttest/utils_conn.py

```diff
@@ -1815,32 +1832,74 @@ def conn_recover(self):
         del self.client_saslconf
         del self.client_hosts
         del self.client_libvirtd_socket
-        # restart libvirtd service on server
-        client_session = self.client_session
         try:
-            libvirtd_service = utils_libvirtd.Libvirtd(session=client_session)
+            sessions = self.session_creator()
+            if not self.run_on_remote:
+                current_session = sessions[0]
+                runner = process.run
+            else:
+                current_session = sessions[1]
+                remote_runner = remote_old.RemoteRunner(session=current_session)
+                runner = remote_runner.run
+            libvirtd_service = utils_libvirtd.Libvirtd(session=current_session)
             if self.libvirt_ver:
                 # For traditional mode, need unmask libvirtd*.socket
                 if traditional_mode == 'yes':
-                    process.run("systemctl unmask libvirtd.socket", shell=True)
-                    process.run("systemctl unmask libvirtd-admin.socket", shell=True)
-                    process.run("systemctl unmask libvirtd-ro.socket", shell=True)
-                    process.run("systemctl unmask libvirtd-tcp.socket", shell=True)
-                    process.run("systemctl unmask libvirtd-tls.socket", shell=True)
+                    runner("systemctl unmask libvirtd.socket")
+                    runner("systemctl unmask libvirtd-admin.socket")
+                    runner("systemctl unmask libvirtd-ro.socket")
+                    runner("systemctl unmask libvirtd-tcp.socket")
+                    runner("systemctl unmask libvirtd-tls.socket")
                 else:
-                    process.run("systemctl daemon-reload", shell=True)
-                    process.run("systemctl stop libvirtd.socket", shell=True)
-                libvirtd_service.stop()
-                process.run("systemctl start libvirtd.socket", shell=True)
+                    runner("systemctl daemon-reload")
+                    runner("systemctl stop libvirtd.socket")
+                runner("systemctl stop libvirtd.service")
+                runner("systemctl start libvirtd.socket")
                 libvirtd_service.start()
             else:
                 libvirtd_service.restart()
+            current_session.close()
         except (remote.LoginError, aexpect.ShellError,
                 process.CmdError) as detail:
             raise ConnServerRestartError(detail)
 
         logging.debug("UNIX connection recover successfully.")
 
+    def session_creator(self):
```

Comment:
I wonder if it's possible to create sessions via `_new_client_session()/_new_server_session()`. It might look simpler... What do you think?

Reply:
Updated by class variables self.client_session and self.server_session as it works on RHEL8.4 as mentioned below.